### DISCUSSION TITLE
Homework 5 Adding a filepath

### DISCUSSION THREAD

**Discussion by MuhammedErkam** [cpp-for-yourself/lectures-and-homeworks#87] [Homeworks Q&A] [answered]
Hi. I am trying to test my classes and I am trying yo give a path like 
`"/home/erkam/Desktop/Projects/CppCourse/cpp-for-yourself-homeworks/homeworks/homework_5/pixelator/pixelator/test_data/test.png"`
But the bot gave me below error:
`Failed to load image data from file: "/home/erkam/Desktop/Projects/CppCourse/cpp-for-yourself-homeworks/homeworks/homework_5/pixelator/pixelator/test_data/test.png"`
How can I fix this problem.
A part of my code :
![Screenshot from 2024-02-19 21-50-23](https://github.com/cpp-for-yourself/lectures-and-homeworks/assets/38918337/afee1e66-8112-4ade-8d6f-7e680fc45888)

**Comment by niosus** [comment, marked as answer]
Hey @MuhammedErkam thanks for the question! This is actually a pretty important one and a very typical situation that happens a lot. The reason for this is that your path is the absolute path that only works for _your_ local system. On your hard drive there is an image under this path, but if you try to run it on a different system, which is what the bot does, the path is going to be different, right? 

To make sure your image is found you will have to have a relative link for the image for now, something like `"../../pixelator/test_data/test.png"`. Think of where your binary is going to be called _from_ and how the path to the image will look like from that place. Does this make any sense?

I actually provide this path in one of the examples in the homework description but might need to stress it more when talking about the test data. Anyway, I think it is an important issue to have as it makes one think about the paths and once you figure it out, you will know this so much better! 🙌

**Reply by MuhammedErkam** [reply]
Thank you so much I tried not to give absolute path also but couldn't write it as yours. I realized my mistake.

**Reply by niosus** [reply]
@MuhammedErkam glad that you figured it out! Take a moment to also think _why_ it works now that it works 😉 And good luck with the rest!

**Reply by MuhammedErkam** [reply]
Hey @niosus I wanna ask you something. I did not realize my mistake totally 😅. Why don't we include screen.hpp file in ftxui like `"../../external/ftxui/include/ftxui/screen/screen.hpp"` in use_ftxui.cpp file. Because binary is inside homework_5/pixelator/build/examples/ folder. I am missing a point and I cannot catch it.

**Reply by niosus** [reply]
@MuhammedErkam no worries, this is a legit question! The reason why we don't use this long path in our source files is that these paths will be fed _into the compiler_. And _it_ will then use it to actually find the needed file. Now the compiler maintains a set of paths where it will look for headers specifically. These can be passed using, for example the `-I` flag that we pass into the compiler binary (which is mostly done through some build system, like CMake in our case). So the header path should be available from one of those paths, not from the file itself. You can see more in [this lecture](https://github.com/cpp-for-yourself/lectures-and-homeworks/blob/main/lectures/headers_and_libraries.md) and it's accompanying video.

That being said, the reason I use a relative path for the image file is that I was lazy to do it the right way. Technically, I would need to copy it into the build folder during the build process so that this path doesn't have to be relative at all. :thinking: I might actually change it some time soon.
